subiquity/server/server.py

```diff
@@ -552,8 +553,93 @@ async def apply_autoinstall_config(self, context):
             await controller.apply_autoinstall_config()
             await controller.configured()
 
+    def _strip_controller_keys(
+        self,
+        autoinstall_config: dict[str, Any],
+    ) -> dict[str, Any]:
+        """Removes all unkown keys from autoinstall config."""
+
+        return_dict: dict = copy.deepcopy(autoinstall_config)
+
+        for controller in self.controllers.instances:
+            return_dict.pop(controller.autoinstall_key, None)
+
+        return return_dict
+
+    @with_context(name="top_level_keys")
+    def _enforce_top_level_keys(
+        self,
+        *,
+        autoinstall_config: dict[str, Any],
+        context: Context,
+    ) -> dict[str, Any]:
+        """Enforces usage of known top-level keys.
+
+        In Autoinstall v1, unknown top-level keys are removed from
+        the config and a cleaned config is returned.
+
+        In Autoinstall v2, unknown top-level keys result in a fatal
+        AutoinstallValidationError
+
+        Only checks for unrecognized keys, doesn't validate them.
+        Requires a version number, so this should be called after
+        validating against the schema.
+
+        """
+
+        server_config: dict[str, Any] = self._strip_controller_keys(autoinstall_config)
+        valid_keys: set[str] = set(self.base_schema["properties"].keys())
+        top_level_keys: set[str] = set(server_config.keys())
+        bad_keys: set[str] = top_level_keys - valid_keys
```

Comment:
And because `server_config` is the unknowns, that means that this calculation of `bad_keys` works because it already has the unknowns stored in `top_level_keys` and we subtract any `valid_keys` (which won't be there) so we get the correct answer but in a bit of an obfuscated way.